compute/cloud-client/src/main/java/compute/custommachinetype/HelperClass.java

```diff
@@ -0,0 +1,269 @@
+              cmt.cpuSeries, cmt.typeLimit.extraMemoryLimit));
+    }
+
+    return null;
+  }
+
+  // Return the custom machine type in the form of a string acceptable by Compute Engine API.
+  public static String returnCustomMachineTypeString(CustomMachineType cmt) {
+    // Check if the requested CPU belongs to E2 series.
+    if (Arrays.asList(CpuSeries.E2_SMALL.name(), CpuSeries.E2_MICRO.name(),
+        CpuSeries.E2_MEDIUM.name()).contains(cmt.cpuSeries)) {
+      return String.format("zones/%s/machineTypes/%s-%s", cmt.zone, cmt.cpuSeries, cmt.memory);
+    }
+
+    // Check if extended memory was requested.
+    if (cmt.memory > cmt.coreCount * cmt.typeLimit.maxMemPerCore) {
+      return String.format("zones/%s/machineTypes/%s-%s-%s-ext", cmt.zone, cmt.cpuSeries,
+          cmt.coreCount,
+          cmt.memory);
+    }
+
+    return String.format("zones/%s/machineTypes/%s-%s-%s", cmt.zone, cmt.cpuSeries, cmt.coreCount,
+        cmt.memory);
+  }
+
+  static class CustomMachineType {
+
+    String zone;
+    String cpuSeries;
+    int memory;
+    int coreCount;
+    TypeLimits typeLimit;
+
+    CustomMachineType(String zone, String cpuSeries, int memory, int coreCount,
+        TypeLimits typeLimit) {
+      this.zone = zone;
+      this.cpuSeries = cpuSeries;
+      this.memory = memory;
+      this.coreCount = coreCount;
+      // Shared machine types (e2-small, e2-medium and e2-micro) always have
+      // 2 vCPUs: https://cloud.google.com/compute/docs/general-purpose-machines#e2_limitations
+      this.typeLimit = typeLimit;
+    }
+
+    @Override
+    public String toString() {
+
+      if (cpuSeries.equalsIgnoreCase(CpuSeries.E2_SMALL.cpuSeries)
+          || cpuSeries.equalsIgnoreCase(CpuSeries.E2_MICRO.cpuSeries)
+          || cpuSeries.equalsIgnoreCase(CpuSeries.E2_MEDIUM.cpuSeries)) {
+        return String.format("zones/%s/machineTypes/%s-%d", zone, cpuSeries, memory);
+      }
+
+      if (memory > typeLimit.maxMemPerCore * coreCount && typeLimit.allowExtraMemory) {
+        return String.format("zones/%s/machineTypes/%s-%d-%d-ext", zone, cpuSeries, coreCount,
+            memory);
+      }
+
+      return String.format("zones/%s/machineTypes/%s-%d-%d", zone, cpuSeries, coreCount, memory);
+    }
+
+    public String shortString() {
+      String cmt = this.toString();
+      return cmt.substring(cmt.lastIndexOf("/") + 1);
+    }
+  }
+
+  // Create a custom machine type.
+  public static CustomMachineType createCustomMachineType(String zone, String cpuSeries, int memory,
+      int coreCount, TypeLimits typeLimit) {
+    if (Arrays.asList(CpuSeries.E2_SMALL.name(), CpuSeries.E2_MICRO.name(),
+        CpuSeries.E2_MEDIUM.name()).contains(cpuSeries)) {
+      coreCount = 2;
+    }
+
+    CustomMachineType cmt = new CustomMachineType(zone, cpuSeries, memory, coreCount, typeLimit);
+
+    try {
+      validate(cmt);
+    } catch (Error e) {
+      // Error in validation.
+      System.out.printf("Error in validation: %s", e);
```

Comment:
If there is an error do we still return the same `cmt` object?

Reply:
Nice catch. Updated it.
Thanks.